lib/rpmrc.c

```diff
@@ -778,23 +778,33 @@ static int get_x86_64_level(void)
 {
     int level = 1;
 
-    unsigned int op_1_ecx = 0, op_80000001_ecx = 0, op_7_ebx = 0, unused;
+    unsigned int op_0_eax = 0, op_1_ecx = 0, op_80000001_ecx = 0, op_7_ebx = 0, unused;
+    cpuid(0, 0, &op_0_eax, &unused, &unused, &unused);
     cpuid(1, 0, &unused, &unused, &op_1_ecx, &unused);
     cpuid(0x80000001, 0, &unused, &unused, &op_80000001_ecx, &unused);
     cpuid(7, 0, &unused, &op_7_ebx, &unused, &unused);
 
+    /* CPUID is unfortunately not quite enough: It indicates whether the hardware supports it,
+     * but software support (kernel/hypervisor) is also needed for register saving/restoring.
+     * For that we can ask the compiler's runtime lib through __builtin_cpu_supports. This is
+     * not usable as a complete replacement for the CPUID code though as not all extensions are
+     * supported (yet). */
+
     const unsigned int op_1_ecx_lv2 = bit_SSE3 | bit_SSSE3 | bit_CMPXCHG16B | bit_SSE4_1 | bit_SSE4_2 | bit_POPCNT;
     if ((op_1_ecx & op_1_ecx_lv2) == op_1_ecx_lv2 && (op_80000001_ecx & bit_LAHF_LM))
 	level = 2;
 
     const unsigned int op_1_ecx_lv3 = bit_FMA | bit_MOVBE | bit_OSXSAVE | bit_AVX | bit_F16C;
     const unsigned int op_7_ebx_lv3 = bit_BMI | bit_AVX2 | bit_BMI2;
-    if (level == 2 && (op_1_ecx & op_1_ecx_lv3) == op_1_ecx_lv3 && (op_7_ebx & op_7_ebx_lv3) == op_7_ebx_lv3
-        && (op_80000001_ecx & bit_LZCNT))
+    if (level == 2 && (op_1_ecx & op_1_ecx_lv3) == op_1_ecx_lv3
+        && op_0_eax >= 7 && (op_7_ebx & op_7_ebx_lv3) == op_7_ebx_lv3
+        && (op_80000001_ecx & bit_LZCNT)
+        && __builtin_cpu_supports("avx2"))
         level = 3;
```

Comment:
This is just cosmetics, but multiline if-conditions should be clearly separated from the following code-block. The && at front helps, but it's better (wrt rpm coding style) to either push the condition continuation to a clearly deeper indentation level, or use a brace on separate line, eg

```
if (level == 2 && (...
                   ...
                   ...))
     level = 3
```

or

```
if (level == 2 && (...
    ...
    ...))
{
    level = 3;
}
```

The latter often ends up being more readable, but depends on the case and individual taste.
Even more often, using meaningfully named helper variables to shorten the condition helps readability even more, but I wouldn't know where to even start here...

Applies to the other if touched in this patch too of course.

Reply:
Fixed